### PR TITLE
[Application Runtime] Support setting probes for the application runtime sidecars

### PR DESCRIPTION
**Pull request by weilerN**
### 📝 Description
Add first-class support for configuring Kubernetes probes (readiness, liveness, startup) on the ApplicationRuntime sidecar.
This includes a clear API for setting and deleting probes, validation of probe types, and automatic enrichment of HTTP probe ports from the internal application port at deploy time.

---

### 🛠️ Changes Made
- Introduced `ProbeType` and `ProbeTimeConfig` enums for consistent probe configuration keys.
- Added `set_probe()` and `delete_probe()` APIs to `ApplicationRuntime` for managing sidecar probes.
- Implemented validation and safe handling of probe types (enum and string-based).
- Added automatic enrichment of HTTP probe ports using `internal_application_port` just before deployment.
- Explicitly blocked probe APIs on non-Application runtimes with clear errors.

---

### ✅ Checklist
- [x] I updated the documentation (if applicable)
- [x] I have tested the changes in this PR
- [ ] I confirmed whether my changes are covered by system tests
  - [ ] If yes, I ran all relevant system tests and ensured they passed before submitting this PR
  - [ ] I updated existing system tests and/or added new ones if needed to cover my changes
- [ ] If I introduced a deprecation:
  - [ ] I followed the [Deprecation Guidelines](./DEPRECATION.md)
  - [ ] I updated the relevant Jira ticket for documentation

---

### 🧪 Testing
- Added unit tests covering all supported scenarios and edge cases
- Dev test: tested that probes changes are propagated properly to the Nuclio side:
  - Set a `sleep` as part of healthcheck function:
  
     ```
       @app.route("/internal")
       def internal():
           sleep(10)
           return "/internal"     
     ```
  - created application runtime and deployment failed (unhealthy):
  `Warning  Unhealthy  3m1s (x867 over 138m)  kubelet  Readiness probe failed: Get "http://10.233.67.57:5000/internal": context deadline exceeded (Client.Timeout exceeded while awaiting headers)`
  
image

  - Configured probes:
  ```
  application.set_probe(type="readiness",
                      initial_delay_seconds=15,
                      period_seconds=10,
                      timeout_seconds=20,
                      http_path="/internal",
                      http_port=5000)
  ```
  - Than deployed again and the function deployed healthy:
  
image

---

### 🔗 References
- Ticket link: https://iguazio.atlassian.net/browse/ML-11708
- Design docs links:
- External links:

---

### 🚨 Breaking Changes?

- [ ] Yes (explain below)
- [x] No

---

### CODE REVIEW

File: mlrun/common/runtimes/constants.py [hunk 1]
```diff
@@ -365,3 +366,23 @@ class NuclioIngressAddTemplatedIngressModes:
...
+    READINESS = "readiness"
+    LIVENESS = "liveness"
+    STARTUP = "startup"
+
+
+PROBE_KEYS = {
+    ProbeType.READINESS.value: "readinessProbe",
+    ProbeType.LIVENESS.value: "livenessProbe",
+    ProbeType.STARTUP.value: "startupProbe",
+}
```
Comment (TomerShor) [inline mlrun/common/runtimes/constants.py]:
```suggestion
class ProbeType(StrEnum):
    READINESS = "readiness"
    LIVENESS = "liveness"
    STARTUP = "startup"

    @property
    def key(self):
        return f"{self.value}Probe"
```

File: mlrun/common/runtimes/constants.py [hunk 2]
```diff
@@ -365,3 +366,23 @@ class NuclioIngressAddTemplatedIngressModes:
...
+
+PROBE_TIMING_INITIAL_DELAY_SECONDS = "initialDelaySeconds"
+PROBE_TIMING_PERIOD_SECONDS = "periodSeconds"
+PROBE_TIMING_TIMEOUT_SECONDS = "timeoutSeconds"
+PROBE_TIMING_FAILURE_THRESHOLD = "failureThreshold"
+PROBE_TIMING_SUCCESS_THRESHOLD = "successThreshold"
```
Comment (TomerShor) [inline mlrun/common/runtimes/constants.py]: This should also be a `StrEnum` class - `ProbeTimeConfig`, instead of just constants.
Then you only need to import the class instead of all separate values, which will make changing it (if needed) much easier.

File: server/py/services/api/api/endpoints/nuclio.py [hunk 3]
```diff
@@ -48,6 +63,178 @@
...
+                error = _validate_http_probe(http_get, probe_type, sidecar_name)
+                if error:
+                    raise mlrun.errors.MLRunBadRequestError(error)
+
+            # Timing parameters validation
+            for field, (validator, desc) in timing_fields.items():
+                if field in probe and not validator(probe[field]):
+                    raise mlrun.errors.MLRunBadRequestError(
+                        f"{probe_type} in sidecar '{sidecar_name}': {field} must be a {desc} integer"
+                    )
```
Comment (TomerShor) [inline server/py/services/api/api/endpoints/nuclio.py]: 1. Private functions should be below Public functions
2. This entire section shouldn't be in this file at all. While relevant for nuclio, it is a very kubernetes thing, so I would put it in `mlrun/k8s_utils.py`. 

Generally, the `endpoints` package should be more of a high level transport layer. It should call the validations but not necessarily implement all of them.
I see that this file doesn't fully adhere to this convention, but it doesn't mean we should continue this.

Comment (TomerShor) [reply server/py/services/api/api/endpoints/nuclio.py]: Also - why is this needed? was this a requirement?
When gets the request from MLRun, it unmarshals the probe configuration into the [Probe Go type](https://pkg.go.dev/k8s.io/api@v0.34.3/core/v1#Probe), so if it doesn't adhere to it - the request will fail. 
So are do we really need to do it in MLRun first?

**Review by TomerShor — CHANGES_REQUESTED** [review]
An initial review

File: mlrun/runtimes/nuclio/application/application.py [hunk 4]
```diff
@@ -294,6 +303,105 @@ def url(self, url):
...
+        :param http_scheme:           "http" or "https" for HTTP probe. Defaults to "http"
+        :param config:                A full dict with the probe configuration
+                                     (used as base, overridden by individual parameters)
+
+        :return: function object (self)
+        """
+        # Validate probe type
+        if not isinstance(type, ProbeType):
+            try:
+                type = ProbeType(type.value if hasattr(type, "value") else type)
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py]: can also be done like here:
https://github.com/mlrun/mlrun/blob/7728707379aaa6efb14e5bc40871df6848b4310c/mlrun/common/schemas/partition.py#L25-L27

File: mlrun/runtimes/nuclio/application/application.py [hunk 5]
```diff
@@ -294,6 +303,105 @@ def url(self, url):
...
+            except (ValueError, AttributeError):
+                raise ValueError(
+                    f"Invalid probe type: {type}. Must be one of: {[p.value for p in ProbeType]}"
+                )
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py]: same here

File: mlrun/runtimes/nuclio/application/application.py [hunk 6]
```diff
@@ -883,3 +992,60 @@ def _sync_api_gateway(self):
...
+
+        Enrichment logic:
+        - Only enriches HTTP probes (httpGet) that don't already have a port specified
+        - If the user explicitly provided http_port in set_probe(), enrichment is skipped
+        - If the user provided a port in the config dict, enrichment is skipped
+        - Enrichment happens just before deployment to ensure the latest internal_application_port
+          value is used, even if it was modified after set_probe() was called
+        """
+        internal_port = None
+        if hasattr(self.spec, "internal_application_port"):
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py]: it is defined as a property https://github.com/mlrun/mlrun/blob/435559a0bc4273fd343312a82723fc05c1f63233/mlrun/runtimes/nuclio/application/application.py#L196-L198, so it always exists, no need to check this again

File: mlrun/runtimes/nuclio/application/application.py [hunk 7]
```diff
@@ -883,3 +992,60 @@ def _sync_api_gateway(self):
...
+                http_get = probe_config.get("httpGet")
+                if http_get and isinstance(http_get, dict) and "port" not in http_get:
+                    if internal_port is None:
+                        raise ValueError(
+                            f"Cannot enrich {probe_type.value} probe: HTTP probe requires a port, "
+                            "but internal_application_port is not set. "
+                            "Please set the internal_application_port or provide http_port in set_probe()."
+                        )
+                    http_get["port"] = internal_port
+                    sidecar[probe_key] = probe_config
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py]: add a debug log with an enriched port value

File: mlrun/runtimes/nuclio/application/application.py [hunk 8]
```diff
@@ -883,3 +992,60 @@ def _sync_api_gateway(self):
...
+            probe_config = sidecar.get(probe_key)
+
+            if probe_config and isinstance(probe_config, dict):
+                http_get = probe_config.get("httpGet")
+                if http_get and isinstance(http_get, dict) and "port" not in http_get:
+                    if internal_port is None:
+                        raise ValueError(
+                            f"Cannot enrich {probe_type.value} probe: HTTP probe requires a port, "
+                            "but internal_application_port is not set. "
+                            "Please set the internal_application_port or provide http_port in set_probe()."
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py:1013]: just fyi - technically it should not be none, because we always enrich it here https://github.com/mlrun/mlrun/blob/a3f2f2f6678de5a462ffa65be0ec2ee295bede73/mlrun/runtimes/nuclio/application/application.py#L162-L165 . So it can be none only if user explicitly set `_internal_application_port` to None

File: server/py/services/api/api/endpoints/nuclio.py (hunk 3 above)
Comment (weilerN) [reply server/py/services/api/api/endpoints/nuclio.py]: @TomerShor After re-reading the requirements, I realized I missed an important part of the validation definition:
```
plus provide the user with an option to supply a dictionary with any parameter that is part of the probe configuration (in which case we will not validate the dictionary and it’s up to the user to make sure the parameters are correct).
```

Given this, the validation I added is not needed on the MLRun server side, so I’m going to remove it.

Comment (weilerN) [reply server/py/services/api/api/endpoints/nuclio.py]: Fixed here - [CR comments 1 - Avoid non required validations](https://github.com/mlrun/mlrun/pull/9043/commits/c2a76bf7e93f934f28d337170f4e30b661155eab)

File: mlrun/runtimes/nuclio/application/application.py [hunk 9]
```diff
@@ -294,6 +299,105 @@ def url(self, url):
...
+        if initial_delay_seconds is not None:
+            probe_config[ProbeTimeConfig.INITIAL_DELAY_SECONDS.value] = (
+                initial_delay_seconds
+            )
+        if period_seconds is not None:
+            probe_config[ProbeTimeConfig.PERIOD_SECONDS.value] = period_seconds
+        if failure_threshold is not None:
+            probe_config[ProbeTimeConfig.FAILURE_THRESHOLD.value] = failure_threshold
+        if timeout_seconds is not None:
+            probe_config[ProbeTimeConfig.TIMEOUT_SECONDS.value] = timeout_seconds
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py]:
```suggestion
probe_config.update({
    config.value: value
    for config, value in {
        ProbeTimeConfig.INITIAL_DELAY_SECONDS: initial_delay_seconds,
        ProbeTimeConfig.PERIOD_SECONDS: period_seconds,
        ProbeTimeConfig.FAILURE_THRESHOLD: failure_threshold,
        ProbeTimeConfig.TIMEOUT_SECONDS: timeout_seconds,
    }.items()
    if value is not None
})
```

**Review by rokatyy — COMMENTED** [review]
some inline comments

File: mlrun/runtimes/nuclio/application/application.py (hunk 8 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py:1013]: @rokatyy Is it possible that `mlrun.mlconf.function.application.default_sidecar_internal_port is None`?
If not than no problem I will remove this check

File: mlrun/runtimes/nuclio/application/application.py (hunk 9 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py]: Fixed here - [CR 1- fixing Kate comments](https://github.com/mlrun/mlrun/pull/9043/commits/a92461e5091f893e3b6a82d91b0c41ddd1a4dbcf)

File: mlrun/runtimes/nuclio/application/application.py (hunk 7 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py]: Fixed here - [CR 1- fixing Kate comments](https://github.com/mlrun/mlrun/pull/9043/commits/a92461e5091f893e3b6a82d91b0c41ddd1a4dbcf)

File: mlrun/runtimes/nuclio/application/application.py (hunk 6 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py]: Fixed here - [CR 1- fixing Kate comments](https://github.com/mlrun/mlrun/pull/9043/commits/a92461e5091f893e3b6a82d91b0c41ddd1a4dbcf)

File: mlrun/runtimes/nuclio/application/application.py (hunk 5 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py]: Fixed here - [CR 1- fixing Kate comments](https://github.com/mlrun/mlrun/pull/9043/commits/a92461e5091f893e3b6a82d91b0c41ddd1a4dbcf)

File: mlrun/runtimes/nuclio/application/application.py (hunk 4 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py]: Fixed here - [CR 1- fixing Kate comments](https://github.com/mlrun/mlrun/pull/9043/commits/a92461e5091f893e3b6a82d91b0c41ddd1a4dbcf)

File: mlrun/runtimes/nuclio/application/application.py (hunk 8 above)
Comment (rokatyy) [reply mlrun/runtimes/nuclio/application/application.py:1013]: @weilerN not really

File: docs/runtimes/application.ipynb [hunk 10]
```diff
@@ -517,6 +517,115 @@
...
+   "metadata": {},
+   "source": [
+    "## Configure Sidecar Kubernetes probes\n",
+    "\n",
+    "You can configure Kubernetes liveness, readiness, and startup probes for the sidecar container using the `set_probe()` and `delete_probe()` methods.\n",
+    "For more information: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes\n",
+    "\n",
+    "### Setting probes\n",
+    "\n",
+    "The `set_probe()` method allows you to configure probes.\n"
```
Comment (TomerShor) [inline docs/runtimes/application.ipynb:532]: Add a note that this overrides any existing probe configuration

File: mlrun/common/runtimes/constants.py [hunk 11]
```diff
@@ -365,3 +366,25 @@ class NuclioIngressAddTemplatedIngressModes:
 class FunctionEnvironmentVariables:
     _env_prefix = "MLRUN_"
     auth_session = f"{_env_prefix}AUTH_SESSION"
+
+
+# Kubernetes probe type keys
```
Comment (TomerShor) [inline mlrun/common/runtimes/constants.py]:
```suggestion
# Kubernetes probe types
```

File: mlrun/runtimes/nuclio/application/application.py [hunk 12]
```diff
@@ -294,6 +299,101 @@ def url(self, url):
...
+                    ProbeTimeConfig.FAILURE_THRESHOLD: failure_threshold,
+                    ProbeTimeConfig.TIMEOUT_SECONDS: timeout_seconds,
+                }.items()
+                if value is not None
+            }
+        )
+
+        # Store probe configuration in the sidecar
+        sidecar = self._set_sidecar(self._get_sidecar_name())
+        sidecar[type.key] = probe_config
```
Comment (TomerShor) [inline mlrun/runtimes/nuclio/application/application.py:369]: What if I don't pass ANY argument? all default to `None`, so will this basically delete the probe configuration, or should we deny this usage?

**Review by TomerShor — CHANGES_REQUESTED** [review]

File: mlrun/runtimes/nuclio/application/application.py (hunk 12 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py:369]: @TomerShor 
- If no arguments are provided (i.e. `application.set_probes()`), the execution will fail because `type` is a mandatory parameter.
- Good question. I’ll raise this with Product. Currently, calling `application.set_probes(type="readiness")` creates a probes sidecar with empty values.

File: mlrun/common/runtimes/constants.py (hunk 11 above)
Comment (weilerN) [reply mlrun/common/runtimes/constants.py]: Fixed here - [CR 2- align function parameters with an user friendly usage](https://github.com/mlrun/mlrun/pull/9043/commits/74071bac9241759e5c2e5a72307ecc8e0b28ae00)

File: docs/runtimes/application.ipynb (hunk 10 above)
Comment (weilerN) [reply docs/runtimes/application.ipynb:532]: Fixed here - [CR 2- align function parameters with an user friendly usage](https://github.com/mlrun/mlrun/pull/9043/commits/74071bac9241759e5c2e5a72307ecc8e0b28ae00)

**Review by TomerShor — APPROVED** [review]
LGTM !
great work

**Review by rokatyy — APPROVED** [review]
LGTM

File: mlrun/runtimes/nuclio/application/application.py (hunk 12 above)
Comment (weilerN) [reply mlrun/runtimes/nuclio/application/application.py:369]: Fixed here - [CR 3 - validate no empty request in client side](https://github.com/mlrun/mlrun/pull/9043/commits/5f850b1334f6bc42386b75bb44dcd77d665c0c6f)

**Comment by TomerShor** [comment]
@weilerN pull latest development, the CI failures should be fixed

**Review by TomerShor — APPROVED** [review]
